teps/0147-dynamic-matrix.md

```diff
@@ -0,0 +1,379 @@
+through a [trigger template params](https://github.com/tektoncd/triggers/blob/main/docs/triggertemplates.md#structure-of-a-triggertemplate).
+Now, running a shared `pipeline` with `matrix` where the values for each instance of fanned out task is dynamically specified
+at the runtime is not possible with the existing `matrix` syntax. In this proposal, we would like to extend `matrix`
+syntax to provide an option to dynamically specify a list of explicit combinations
+
+### Goals
+
+- Be able to author a sharable pipeline with `matrix` such that a task can fan out based on the explicit combinations
+which are specified dynamically.
+
+### Non-Goals
+
+- Matrix explicit combination mechanism only supports a `param` of type `string` and not adding support for any
+other type.
+
+### Use Cases
+
+Let's revisit the example of build pipeline from [TEP-0118](0118-matrix-with-explicit-combinations-of-parameters.md#define-explicit-combinations-in-the-matrix).
+A pipeline `pipeline-to-build-images-from-a-single-repo` has three explicit combinations defined in the pipeline such
+that `kaniko` can fan out and create three separate `taskRun`s. The values for each combination can be furthered
+dynamically specified using the `params` as seen in the example in the [summary](#summary).
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: pipeline-to-build-images-from-a-single-repo
+spec:
+  workspaces:
+    - name: shared-workspace
+  tasks:
+    - ...
+    - name: kaniko-build
+      taskRef:
+        name: kaniko
+      workspaces:
+        - name: source
+          workspace: shared-workspace
+      matrix:
+        include:
+          - name: build-1
+            params:
+              - name: IMAGE
+                value: "image-1"
+              - name: DOCKERFILE
+                value: "path/to/Dockerfile1"
+          - name: build-2
+            params:
+              - name: IMAGE
+                value: "image-2"
+              - name: DOCKERFILE
+                value: "path/to/Dockerfile2"
+          - name: build-3
+            params:
+              - name: IMAGE
+                value: "image-3"
+              - name: DOCKERFILE
+                value: "path/to/Dockerfile3"
+    - ...
+```
+
+This `pipeline` is part of a catalog and shared across multiple teams. Now, the teams are building their own applications
+using this `pipeline` and different teams have different number of images to built as part of their application. This way
+of specifying `matrix` combinations in a `pipeline` is constant and can not be utilized for an application with
+different number of images.
+
+## Proposal
+
+We propose adding a field - `strategy` of type `string` - within the `matrix` section. This allows pipeline authors to
+maintain a single pipeline in a catalog and allow the users to specify the list of explicit combinations for each
+instance dynamically through `pipelineRun` or a trigger template or a task result.
```

Comment:
yup, it could be confusing to the users directly coming in from Github actions but we have designed it such that it align with `pipelineTask` struct. 

We have introduced `matrix` at the `taskRef` level in the `pipelineTask` struct. All `matrix` related fields are part of `pipelineTask.matrix`.

Supporting an expression language is listed as a future action item. As a project, Tekton pipelines can decide which all expressions languages we would like to support and design accordingly. We have a `CEL` being adopted in `when` expressions. We can support it here as well in `matrix.strategy` if we have a use case.

Reply:
I'm just wondering if we could have a better naming than `strategy`? The new field is designed to accept dynamic matrix to pass in. I don't see a clear connection between the word  `strategy` and its purpose.